msyslog-v1.08a+smac/src/man/om_classic.8
.\"	$CoreSDI: om_classic.8,v 1.1.2.4 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2000, 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Apr 06, 2001
.Dt IM_CLASSIC 8
.Os Core-SDI
.Sh NAME
.Nm classic output module
.Nd
.Xr syslogd 8
output module for output to files, users or other hosts (UDP)
.Sh SYNOPSIS
.Nm classic
.Ar filename | user{,user} | * | @hostname
.Sh DESCRIPTION
If there is no module mentioned in a
.Xr syslog.conf 5
line,
.Xr om_classic 8
is used.
.Bl -tag -width Ds
.It Ar filename
Appends logs to
.Pa filename .
Note that
.Xr om_classic 8
will not create the logfile for you.
.It Ar user{,user}
Writes the log messages to the tty of the listed users.
.It Ar *
Writes the logs to all logged in users.
.It Ar @hostname
Sends logs with UDP to the "syslog" port on
.Ar hostname .
.El
.Pp
.Sh EXAMPLES
.Bd -literal -offset indent
# logs to a file
*.*		/var/log/all
# the same, with module syntax
*.*		%classic /var/log/all
.Pp
# to all logged in users
*.emerg		%classic *
.Pp
# to another machine by means of UDP
*.notice		%classic @loghost.domain.com
.Pp
# to the terminal where root is logged in.
*.alert		%classic root
.Ed
.Pp
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr syslog.conf 5 ,
.Xr om_mysql 8 ,
.Xr om_peo 8 ,
.Xr om_pgsql 8 ,
.Xr om_regex 8 ,
.Xr om_tcp 8 ,
.Xr om_udp 8 ,
.Xr peochk 8 ,
.Xr syslogd 8
.Sh BUGS
.Bl -bullet
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/om_classic.c
/*	$CoreSDI: om_classic.c,v 1.31.2.8.2.4.4.22 2001/11/21 06:31:03 alejo Exp $	*/
/*
 * Copyright (c) 1983, 1988, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 *  om_classic -- classic behaviour module
 *
 * Author: Alejo Sanchez for Core-SDI SA
 *         from syslogd.c Eric Allman  and Ralph Campbell
 *
 */

#include "config.h"

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <sys/param.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include <sys/uio.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <utmp.h>
#include <netdb.h>
/* if _PATH_UTMP isn't defined, define it here... */
#ifndef _PATH_UTMP
# ifdef UTMP_FILE
#  define _PATH_UTMP UTMP_FILE
# else  /* if UTMP_FILE */
#  define _PATH_UTMP "/var/adm/utmp"
# endif /* if UTMP_FILE */
#endif


#include "../modules.h"
#include "../syslogd.h"

#define TTYMSGTIME	1		/* timeout passed to ttymsg */

/* values for f_type */
#define F_UNUSED	0		/* unused entry */
#define F_FILE		1		/* regular file */
#define F_TTY		2		/* terminal */
#define F_CONSOLE	3		/* console terminal */
#define F_FORW		4		/* remote machine */
#define F_USERS		5		/* list of users */
#define F_WALL		6		/* everyone logged on */
#define F_PIPE		7		/* named pipe */

/* names for f_types */
char    *TypeNames[] = { "UNUSED", "FILE", "TTY", "CONSOLE",
	    "FORW", "USERS", "WALL", "PIPE", NULL};

struct om_classic_ctx {
	int	fd;
	union {
		char    f_uname[MAXUNAMES][UT_NAMESIZE+1];
		struct {
			char    f_hname[MAXHOSTNAMELEN];
			struct sockaddr		f_addr;
		} f_forw;	/* forwarding address */
		char    f_fname[MAXPATHLEN];
	} f_un;
	int	f_type;		 /* entry type, see below */
};

void wallmsg (struct filed *, struct iovec *, struct om_classic_ctx *c);
char *ttymsg(struct iovec *, int , char *, int);
struct sockaddr	*resolv_name(char *, char *, char *, size_t *);


/*
 * Write to file, tty, user and network udp
 */

int
om_classic_write(struct filed *f, int flags, struct m_msg *m, void *ctx)
{
	struct iovec iov[6];
	struct iovec *v;
	struct om_classic_ctx *c;
	int l;
	char line[MAXLINE + 1], greetings[500], time_buf[16];
	time_t now;

	if (m == NULL || m->msg == NULL || !strcmp(m->msg, "")) {
		dprintf(MSYSLOG_INFORMATIVE, "om_classic_write: no message!");
		return (-1);
	}

	c = (struct om_classic_ctx *) ctx;

	/* prepare buffers for writing */
	v = iov;
	if (c->f_type == F_WALL) {
		v->iov_base = greetings;
		v->iov_len = snprintf(greetings, sizeof(greetings),
		    "\r\n\7Message from syslogd@%s at %.24s ...\r\n",
		    f->f_prevhost, ctime(&now));
		if (v->iov_len >= sizeof(greetings))
			v->iov_len = sizeof(greetings) - 1;
		v++;
		v->iov_base = "";
		v->iov_len = 0;
		v++;
	} else {
		strftime(time_buf, sizeof(time_buf), "%b %e %H:%M:%S", &f->f_tm);
		v->iov_base = time_buf;
		v->iov_len = 15;
		v++;
		v->iov_base = " ";
		v->iov_len = 1;
		v++;
	}
	v->iov_base = f->f_prevhost;
	v->iov_len = strlen(v->iov_base);
	v++;
	v->iov_base = " ";
	v->iov_len = 1;
	v++;

	v->iov_base = m->msg;
	v->iov_len = strlen(m->msg);
	v++;

	dprintf(MSYSLOG_INFORMATIVE, "Logging to %s", TypeNames[c->f_type]);

	switch (c->f_type) {
	case F_UNUSED:
		dprintf(MSYSLOG_INFORMATIVE, "\n");
		break;
	
	case F_FORW:
		if (c->fd < 0) {
			dprintf(MSYSLOG_SERIOUS, "om_classic: write: "
			    "can't forward message, socket down\n");
			return(-1);
		}

		dprintf(MSYSLOG_INFORMATIVE, " %s\n", c->f_un.f_forw.f_hname);
		l = snprintf(line, sizeof(line), "<%d>%.15s %s", f->f_prevpri,
		    (char *) iov[0].iov_base, (char *) iov[4].iov_base);

		if (sendto(c->fd, line, l, 0, &c->f_un.f_forw.f_addr,
#ifdef AF_INET6
		    c->f_un.f_forw.f_addr.sa_family == AF_INET6 ?
		    sizeof(struct sockaddr_in6) :
#endif
		    sizeof(struct sockaddr_in)) != l) {
			c->f_type = F_UNUSED;
			dprintf(MSYSLOG_WARNING, "om_classic: error on sendto()");
		}

		break;

	case F_CONSOLE:
		if (flags & IGN_CONS) {
			dprintf(MSYSLOG_INFORMATIVE, " (ignored)\n");
			break;
		}
		/* FALLTHROUGH */

	case F_TTY:
	case F_PIPE:
	case F_FILE:
		dprintf(MSYSLOG_INFORMATIVE, " %s\n", c->f_un.f_fname);
		if (c->f_type != F_FILE) {
			v->iov_base = "\r\n";
			v->iov_len = 2;
		} else {
			v->iov_base = "\n";
			v->iov_len = 1;
		}
		again:
		if (writev(c->fd, iov, 6) < 0) {
			int e = errno;

			/* from sysklogd */
			/* If a named pipe is full, just ignore */
			if (c->f_type == F_PIPE && e == EAGAIN)
				break;

			close(c->fd);

			/*
			 * Check for errors on TTY's due to loss of tty
			 */
			if ((e == EIO || e == EBADF) && c->f_type != F_FILE) {
				c->fd = open(c->f_un.f_fname,
				    O_WRONLY|O_APPEND, 0);
				if (c->fd < 0) {
					c->f_type = F_UNUSED;
					dprintf(MSYSLOG_WARNING, "om_classic: "
					    "error on %s", c->f_un.f_fname);
				} else
					goto again;
			} else {
				c->f_type = F_UNUSED;
				c->fd = -1;
				errno = e;
				dprintf(MSYSLOG_WARNING, "om_classic: error "
				    "on %s", c->f_un.f_fname);
			}
		} else if (flags & SYNC_FILE)
			fsync(c->fd);
		break;

	case F_USERS:
	case F_WALL:
		dprintf(MSYSLOG_INFORMATIVE, "\n");
		v->iov_base = "\r\n";
		v->iov_len = 2;
		wallmsg(f, iov, c);
		break;
	}
	f->f_prevcount = 0;

	return (1);
}


/*
 *  INIT -- Initialize om_classic
 *
 *  taken mostly from syslogd's cfline
 */
int
om_classic_init(int argc, char **argv, struct filed *f, char *prog, void **ctx,
    char **status)
{
	struct sockaddr *sa;
	struct  om_classic_ctx *c;
	size_t salen;
	int i, statbuf_len;
	char *p, *q, statbuf[1024];

	dprintf(MSYSLOG_INFORMATIVE, "om_classic_init: Entering\n");

	/* accepts "%classic /file" or "%classic -t TYPE /file" */
	if ( (argc != 2 && argc != 4) || argv == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_classic_init: incorrect "
		    "parameters %d args [%s %s %s %s]\n", argc,
		    argc > 0? argv[1] : "",
		    argc > 1? argv[2] : "",
		    argc > 2? argv[3] : "",
		    argc > 3? argv[4] : "");
		return (-1);
	}

	if ((*ctx = (void *) calloc(1, sizeof(struct om_classic_ctx))) == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_classic_init: cannot allocate "
		    "context\n");
		return (-1);
	}

	c = (struct om_classic_ctx *) *ctx;

	if (argc > 2) {

		if (strncmp(argv[1], "-t", 2)) {
			dprintf(MSYSLOG_SERIOUS, "om_classic_init: incorrect" 
			    " parameter %s, should be '-t'\n", argv[1]);
			free(*ctx);
			*ctx = NULL;
			return (-1);
		}

		/* look for entry # in table */
		for (i = 0; TypeNames[i] && strncmp(TypeNames[i], argv[2],
		    strlen(argv[2])); i++);
		if (TypeNames[i] == NULL) {
			dprintf(MSYSLOG_SERIOUS, "om_classic_init: couldn't" 
			    " determine type %s\n", argv[2]);
			free(*ctx);
			*ctx = NULL;
			return (-1);
		}

		c->f_type = i;
		p = argv[3];

	} else
		/* regular config line, no type */
		p = argv[1];

	switch (*p) {
	case '@':
		c->fd = socket(AF_INET, SOCK_DGRAM, 0);

		strncpy(c->f_un.f_forw.f_hname, ++p,
		    sizeof(c->f_un.f_forw.f_hname) - 1);
		c->f_un.f_forw.f_hname[sizeof(c->f_un.f_forw.f_hname) - 1]
		    = '\0';

		if ((sa = resolv_name(c->f_un.f_forw.f_hname, "syslog", "udp",
		    &salen)) == NULL) {
			dprintf(MSYSLOG_SERIOUS, "om_classic: error resolving "
			    "host %s\n", c->f_un.f_forw.f_hname);
			break;
		}

		memmove(&c->f_un.f_forw.f_addr, sa, salen);
		free(sa);
		c->f_type = F_FORW;
		snprintf(statbuf, sizeof(statbuf), "om_classic: "
		    "forwarding messages through UDP to host %s",
		    c->f_un.f_forw.f_hname);
		break;

	case '-':  /* ignore this, we do it by default */
		p++;
	case '|':  /* from sysklogd */
	case '/':
		strncpy(c->f_un.f_fname, p, sizeof c->f_un.f_fname);
		c->f_un.f_fname[sizeof (c->f_un.f_fname) - 1] = 0;
		if ( *p == '|' ) {
			c->fd = open(++p, O_RDWR|O_NONBLOCK);
			c->f_type = F_PIPE;
		} else {
			c->fd = open(p, O_WRONLY|O_APPEND, 0);
			c->f_type = F_FILE;
		}

		if (c->fd < 0) {
			dprintf(MSYSLOG_CRITICAL, "om_classic_init: error "
			    "opening log file: %s\n", p);
			free(*ctx);
			*ctx = NULL;
			return (-1);
		}

		if (!c->f_type) {
			if (isatty(c->fd))
				c->f_type = F_TTY;
			else
				c->f_type = F_FILE;
		}

		snprintf(statbuf, sizeof(statbuf), "om_classic: "
		    "saving messages to file %s", c->f_un.f_fname);
		break;

	case '*':
		c->f_type = F_WALL;
		snprintf(statbuf, sizeof(statbuf), "om_classic: sending "
		    "messages to all logged users");

	default:
		for (i = 0; i < MAXUNAMES && *p; i++) {
			for (q = p; *q && *q != ','; )
				q++;
			(void)strncpy(c->f_un.f_uname[i], p, UT_NAMESIZE);
			if ((q - p) > UT_NAMESIZE)
				c->f_un.f_uname[i][UT_NAMESIZE] = '\0';
			else
				c->f_un.f_uname[i][q - p] = '\0';
			while (*q == ',' || *q == ' ')
				q++;
			p = q;
		}
		c->f_type = F_USERS;
		statbuf_len = snprintf(statbuf, sizeof(statbuf),
		    "om_classic: forwarding messages to users:");
		for (i = 0; i < MAXUNAMES &&
		    c->f_un.f_uname[i][0] != '\0'; i++) {
			statbuf_len += snprintf(statbuf,
			    sizeof(statbuf) - statbuf_len, " %s",
			    c->f_un.f_uname[i]);
		}
		break;
	}

	*status = strdup(statbuf);

	return (1);
}

int
om_classic_close(struct filed *f, void *ctx)
{
	struct om_classic_ctx *c;

	c = (struct om_classic_ctx *) ctx;

	if (c->fd > -1)
		close(c->fd);

	return (0);
}

int
om_classic_flush(struct filed *f, void *ctx)
{
	/* flush any pending output */
	if (f->f_prevcount)
		om_classic_write(f, 0, NULL, NULL);

	return (1);
}

/*
 *  WALLMSG -- Write a message to the world at large
 *
 *	Write the specified message to either the entire
 *	world, or a list of approved users.
 */
void
wallmsg( struct filed *f, struct iovec *iov, struct om_classic_ctx *c)
{
	static int reenter;			/* avoid calling ourselves */
	FILE *uf;
	struct utmp ut;
	int i;
	char *p;
	char line[sizeof(ut.ut_line) + 1];

	if (reenter++)
		return;
	if ( (uf = fopen(_PATH_UTMP, "r")) == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_classic: error opening "
		    "%s\n", _PATH_UTMP);
		reenter = 0;
		return;
	}
	/* NOSTRICT */
	while (fread(&ut, sizeof(ut), 1, uf) == 1) {

#ifndef __linux__
		if (ut.ut_name[0] == '\0')
#else
		if ((ut.ut_type != USER_PROCESS && ut.ut_type != LOGIN_PROCESS) ||
		    ut.ut_line[0] == ':' /* linux logs users that are not logged in (?!) */)
#endif
			continue;

		strncpy(line, ut.ut_line, sizeof(ut.ut_line));
		line[sizeof(ut.ut_line)] = '\0';
		if (c->f_type == F_WALL) {
			if ((p = ttymsg(iov, 6, line, TTYMSGTIME)) != NULL) {
				errno = 0;	/* already in msg */
				dprintf(MSYSLOG_SERIOUS, "om_classic: error "
				    "%s\n", p);
			}
			continue;
		}
		/* should we send the message to this user? */
		for (i = 0; i < MAXUNAMES; i++) {
			if (!c->f_un.f_uname[i][0])
				break;
			if (!strncmp(c->f_un.f_uname[i], ut.ut_name,
			    UT_NAMESIZE)) {
				if ((p = ttymsg(iov, 6, line, TTYMSGTIME))
								!= NULL) {
					errno = 0;	/* already in msg */
					dprintf(MSYSLOG_SERIOUS, "om_classic: error "
					    "%s\n", p);
				}
				break;
			}
		}
	}
	fclose(uf);
	reenter = 0;
}
